README.md
<p align="center">
  <img
    src="http://res.cloudinary.com/vidsy/image/upload/v1503160820/CoZ_Icon_DARKBLUE_200x178px_oq0gxm.png"
    width="125px;">
</p>

<h1 align="center">neo-python-rpc</h1>

<p align="center">A lightweight Python RPC Client for the NEO Blockchain</p>
<div align="center">

<a href="https://pypi.python.org/pypi/neo-python-rpc" rel="nofollow"><img src="https://img.shields.io/pypi/v/neo-python-rpc.svg">
</a>
<a href="https://travis-ci.org/CityOfZion/neo-python-rpc" rel="nofollow"><img src="https://img.shields.io/travis/CityOfZion/neo-python-rpc.svg">
</a>
<a href="https://neo-python-rpc.readthedocs.io/en/latest/?badge=latest" rel="nofollow"><img src="https://readthedocs.org/projects/neo-python-rpc/badge/?version=latest"></a>
<a href="https://coveralls.io/github/CityOfZion/neo-python-rpc?branch=master" rel="nofollow"><img src="https://coveralls.io/repos/github/CityOfZion/neo-python-rpc/badge.svg?branch=master"></a>
</div>

<ul>
<li>Free software: MIT license</li>
<li>Documentation: <a href="https://neo-python-rpc.readthedocs.io" rel="nofollow">https://neo-python-rpc.readthedocs.io</a>.</li>
</ul>


## Install

`pip install neo-python-rpc`



### Basic Usage

- Get Height of blockchain
    ```
    >>> from neorpc.Client import RPCClient
    >>> client = RPCClient()
    >>> blockchain_height = client.get_height()
    >>> blockchain_height
    769332
    ```
    
- Get a block
    ```
    >>> block = client.get_block(123456)
    >>> block
    {
        "nextconsensus": "AdyQbbn6ENjqWDa5JNYMwN3ikNcA4JeZdk",
        "size": 686,
        "previousblockhash": "0xdbee7804eb1f648e640401a864ed29bbc787d7f9d36b6f4be0a3ffd693723d5e",
        "script": {
            "invocation": "4012bbee0bac21b6ec5c4b0cda3aa28f3aa51e3b4a6c230908370adc8758fc3b051a865539bdca0cc6b450ff4fa18d3295f5c5a05a862b8aa9baff3e69011cb0f640ca5e3e5b3399a2014aa96575
    e2b0c3d1239d41ac05a92ec430ab0e0388f5e76e30062dd9aecca3007993e8bde7a00b70cdbef2fbb641164b9fba5f6ed724dc2e40458bcdf880cfef8d94aef714274ff4c258a13f541d2f8f678520bb5414704662cc5b9e714
    f2e42c94c2a61f8e0b3fb2ec8cc6e9ef2e0670056082370e9331e0f40f0a005fbb338e69465c6e0760ac04c2d61bdbec9922f10b8936d277609dd0f6cab1d7ee944902e0af2ba4ac6183a30158c1aa96c77262070161d24e1c5
    530842403cc25ba0ce9ca460d3bf6108a48c676f131076c1542173278cace43a69b7038d4f2f5b401e8bdea2e10c9c68598d9f530b6618b3165a7c791442cd6d093b6aa2",
            "verification": "55210209e7fd41dfb5c2f8dc72eb30358ac100ea8c72da18847befe06eade68cebfcb9210327da12b5c40200e9f65569476bbff2218da4f32548ff43b6387ec1416a231ee821034ff5ceeac41a
    cf22cd5ed2da17a6df4dd8358fcb2bfb1a43208ad0feaab2746b21026ce35b29147ad09e4afe4ec4a7319095f08198fa8babbe3c56e970b143528d2221038dddc06ce687677a53d54f096d2591ba2302068cf123c1f2d75c2dd
    dc542557921039dafd8571a641058ccc832c5e2111ea39b09c0bde36050914384f7a48bce9bf92102d02b1873a0863cd042cc717da31cea0d7cf9db32b74d4c72c01b0011503e2e2257ae"
        },
        "tx": [
            {
                "vin": [],
                "size": 10,
                "type": "MinerTransaction",
                "txid": "0x0981c5e57997af0ce18d032bc9969220308dce8ec643844398649b6732ad91fe",
                "attributes": [],
                "net_fee": "0",
                "vout": [],
                "nonce": 1808064137,
                "version": 0,
                "scripts": [],
                "sys_fee": "0"
            }
        ],
        "nextblockhash": "0xed2cf18a0b15bb3b6a244abd2de1227afde61d1b0a9cbc582e587c18b3c3a00e",
        "confirmations": 647426,
        "merkleroot": "0x0981c5e57997af0ce18d032bc9969220308dce8ec643844398649b6732ad91fe",
        "hash": "0x10d1473c9fa133b28ea4f470e58b97bfe1222e102b14fbc50d3ade59eb8e3b74",
        "version": 0,
        "time": 1496944720,
        "nonce": "c8f9672b6bc4de89",
        "index": 123456
    }
    ```
    
    
    

HISTORY.rst
=======
History
=======

0.1.2 (2017-11-12)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for neorpc-python-rpc."""


tests/test_client.py
from unittest import TestCase
from neorpc.Settings import SettingsHolder
from neorpc.Client import RPCClient,RPCEndpoint
import binascii


class RPCClientTestCase(TestCase):

    sample_addr = 'AXjaFSP23Jkbe6Pk9pPGT6NBDs1HVdqaXK'

    sample_asset = 'c56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b'

    def test_client(self):

        client = RPCClient()

        self.assertIsNotNone(client.endpoints)

        self.assertGreater(len(client.endpoints), 0)

        self.assertIsInstance(client.default_endpoint, RPCEndpoint)

        self.assertEqual(client.default_endpoint.height, None)

    def test_settings(self):
        settings = SettingsHolder()
        settings.setup_mainnet()

        client = RPCClient(config=settings)

        self.assertIsNotNone(client.endpoints)

    def test_client_setup(self):

        client = RPCClient(setup=True)

        self.assertIsNotNone(client.endpoints)

        self.assertGreater(len(client.endpoints), 0)

        self.assertIsInstance(client.default_endpoint, RPCEndpoint)

        self.assertIsNotNone(client.default_endpoint.height)

        self.assertEqual(client.default_endpoint.status, 200)

    def test_height(self):
        client = RPCClient()

        response = client.get_height()

        height = int(response)

        self.assertGreaterEqual(height, 0)

    def test_account_state(self):

        client = RPCClient()

        account = client.get_account(self.sample_addr)

        script_hash = bytearray(binascii.unhexlify(account['script_hash'][2:]))
        script_hash.reverse()

        self.assertEqual(len(script_hash), 20)

        balances = account['balances']

        self.assertIsNotNone(balances)
        self.assertGreater(len(balances), 0)

    def test_asset_state(self):

        client = RPCClient()

        asset = client.get_asset(self.sample_asset)

        self.assertEqual(asset['type'], 'GoverningToken')

        self.assertEqual(int(asset['available']), 100000000)

        self.assertEqual(int(asset['precision']), 0)

    def test_best_blockhash(self):

        client = RPCClient()

        hash = bytearray(binascii.unhexlify(client.get_best_blockhash()[2:]))
        hash.reverse()

        self.assertEqual(len(hash), 32)

    def test_get_block_json(self):

        client = RPCClient()

        height = 12344
        blockjson1 = client.get_block(height)

        blockhash = blockjson1['hash'][2:]

        self.assertEqual(blockhash, '1e67372c158a4cfbb17b9ad3aaae77001a4247a00318e354c62e53b56af4006f')

        blockjson2 = client.get_block(blockhash)

        self.assertEqual(blockjson1, blockjson2)

        self.assertEqual(height, blockjson1['index'])

    def test_getblockhash(self):

        client = RPCClient()

        height = 12344

        hash = client.get_block_hash(height)

        self.assertEqual(hash[2:], '1e67372c158a4cfbb17b9ad3aaae77001a4247a00318e354c62e53b56af4006f')

    def test_sysfee(self):

        client = RPCClient()

        height = 740324

        fee = int(client.get_block_sysfee(height))

        self.assertEqual(fee, 469014)

    def test_contract_state(self):

        client = RPCClient()

        contract_hash = 'f8d448b227991cf07cb96a6f9c0322437f1599b9'

        contract = client.get_contract_state(contract_hash)

        hash = contract['hash'][2:]

        self.assertEqual(hash, contract_hash)

    def test_connection_count(self):

        client = RPCClient()

        connection_count = int(client.get_connection_count())

        self.assertGreater(connection_count, 0)

    def test_mempool(self):

        client = RPCClient()

        mempool = client.get_raw_mempool()

        self.assertIsInstance(mempool, list)

    def test_tx_json(self):

        client = RPCClient()

        hash = '58c634f81fbd4ae2733d7e3930a9849021840fc19dc6af064d6f2812a333f91d'
        tx = client.get_transaction(hash)

        self.assertEqual(tx['blocktime'], 1510283768)

        self.assertEqual(tx['txid'][2:], hash)

    def test_get_storage(self):

        client = RPCClient()

        contract_hash = 'd7678dd97c000be3f33e9362e673101bac4ca654'
        key = 'totalSupply'

        storage = client.get_storage(contract_hash, key)

        self.assertIsInstance(storage, bytearray)

        intval = int.from_bytes(storage, 'little')

        self.assertEqual(intval, 196800000000000)

    def test_tx_out(self):

        client = RPCClient()

        tx_hash = '58c634f81fbd4ae2733d7e3930a9849021840fc19dc6af064d6f2812a333f91d'
        index = 0
        txout = client.get_tx_out(tx_hash, index)

        self.assertEqual(txout['asset'][2:], '602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7')
        self.assertEqual(txout['n'], index)
        self.assertEqual(txout['address'], self.sample_addr)

    def test_invoke(self):

        client = RPCClient()

        contract_hash = 'd7678dd97c000be3f33e9362e673101bac4ca654'
        params = [{'type': 7, 'value': 'symbol'}, {'type': 16, 'value': []}]

        result = client.invoke_contract(contract_hash, params)

        self.assertEqual(result['state'], 'HALT, BREAK')
        invoke_result = result['stack']
        self.assertEqual(len(invoke_result), 1)

        stack_item = invoke_result[0]

        self.assertEqual(stack_item['type'], 'ByteArray')
        self.assertEqual(binascii.hexlify('LWTF'.encode('utf-8')).decode('utf-8'), stack_item['value'])

    def test_invoke_fn(self):

        client = RPCClient()

        contract_hash = 'd7678dd97c000be3f33e9362e673101bac4ca654'

        result = client.invoke_contract_fn(contract_hash, 'name')

        self.assertEqual(result['state'], 'HALT, BREAK')
        invoke_result = result['stack']
        self.assertEqual(len(invoke_result), 1)

        invoke_result = result['stack']
        self.assertEqual(len(invoke_result), 1)

        stack_item = invoke_result[0]

        self.assertEqual(stack_item['type'], 'ByteArray')

        val = binascii.unhexlify(stack_item['value'].encode('utf-8')).decode('utf-8')

        self.assertEqual(val, 'LOCALTOKEN')

    def test_invoke_script(self):

        client = RPCClient()

        script = '00c10b746f74616c537570706c796754a64cac1b1073e662933ef3e30b007cd98d67d7'

        result = client.invoke_script(script)

        self.assertEqual(result['state'], 'HALT, BREAK')
        invoke_result = result['stack']
        self.assertEqual(len(invoke_result), 1)

        invoke_result = result['stack']
        self.assertEqual(len(invoke_result), 1)

        stack_item = invoke_result[0]

        self.assertEqual(stack_item['type'], 'ByteArray')

        val = int.from_bytes(binascii.unhexlify(stack_item['value'].encode('utf-8')), 'little')

        self.assertEqual(val, 196800000000000)

    def test_send_raw_tx(self):

        client = RPCClient()

        raw_tx = '80000120d8edd2df8d6907caacd4af8872a596cb75c5829d015ce72895ce376d12def9a780ba502ae28ad7a4b7fbcf6baa4856edb537417d2a0000029b7cffdaa674beae0f930ebe6085af9093e5fe56b34a5c220ccdcf6efc336fc500a3e11100000000d8edd2df8d6907caacd4af8872a596cb75c5829d9b7cffdaa674beae0f930ebe6085af9093e5fe56b34a5c220ccdcf6efc336fc500bbeea000000000d8edd2df8d6907caacd4af8872a596cb75c5829d01414044dfd2b360e548607ece3d453173079233040c2484a99671a7346a8ca16969245b946bfa4c13125f4c931b0cbab216e0d241d908f37ad96abb776890832a3a4b2321025de86902ed42aca7246207a70869b22253aeb7cc84a4cb5eee3773fd78b3f339ac'

        # this will result in a false, since this tx has already been made
        # but, if it were badly formmated, it would be an error
        # so we are testing if we get back a false
        result = client.send_raw_tx(raw_tx)

        self.assertEqual(result, False)

    def test_validate_addr(self):

        client = RPCClient()

        result = client.validate_addr(self.sample_addr)

        self.assertEqual(result['isvalid'], True)
        self.assertEqual(result['address'], self.sample_addr)

        bad_addr = 'BXjaFSP23Jkbe6Pk9pPGT6NBDs1HVdqaXK'

        result = client.validate_addr(bad_addr)

        self.assertEqual(result['isvalid'], False)

    def test_get_peers(self):

        client = RPCClient()

        result = client.get_peers()

        self.assertIsNotNone(result)
#        self.assertIn('connected', result)
#        self.assertIn('unconnected', result)


# Not all endpoints currently implement this method
#
#    def test_get_version(self):
#
#        client = RPCClient()
#
#        version = client.get_version()
#
#        print("versin: %s " % version)


neorpc/Settings.py
"""

Sets up your RPC endpoints


"""


class SettingsHolder:
    """
    This class holds all the settings. Needs to be setup with one of the
    `setup` methods before using it.
    """
    RPC_LIST = None

    # Setup methods
    def setup(self, addr_list):
        """ Load settings from a JSON config file """
        self.RPC_LIST = addr_list

    def setup_mainnet(self):
        """ Load settings from the mainnet JSON config file """
        self.setup(
            [
                "http://13.59.52.94:8080",
                "http://18.220.214.143:8080",
                "http://13.58.198.112:8080",
                "http://13.59.14.206:8080",
                "http://18.216.9.7:8080"
            ]
        )

    def setup_testnet(self):
        self.setup(
            [
                "http://18.221.221.195:8880",
                "http://18.221.139.152:8880",
                "http://52.15.48.60:8880",
                "http://18.221.0.152:8880",
                "http://52.14.184.44:8880"
            ]
        )

    def setup_privnet(self):
        """ Load settings from the privnet JSON config file """
        self.setup(
            [
                "127.0.0.1:20332"
            ]
        )


# Settings instance used by external modules
settings = SettingsHolder()

# Load testnet settings as default
settings.setup_testnet()


neorpc/Client.py
from neorpc.Settings import settings as rpc_settings
import requests
import binascii


class RPCClient():

    id_counter = 0

    _settings = rpc_settings
    _addr_list = None

    @property
    def endpoints(self):
        return self._addr_list

    @property
    def default_endpoint(self):
        self._addr_list.sort()
        return self._addr_list[0]

    def get_account(self, address, id=None, endpoint=None):
        """
        Look up an account on the blockchain.  Sample output:

        Args:
            address: (str) address to lookup ( in format 'AXjaFSP23Jkbe6Pk9pPGT6NBDs1HVdqaXK')
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use

        Returns:
            json object of the result or the error encountered in the RPC call

        """
        return self._call_endpoint(GET_ACCOUNT_STATE, params=[address], id=id, endpoint=endpoint)

    def get_height(self, id=None, endpoint=None):
        """
        Get the current height of the blockchain
        Args:
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use

        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_BLOCK_COUNT, id=id, endpoint=endpoint)

    def get_asset(self, asset_hash, id=None, endpoint=None):
        """
        Get an asset by its hash
        Args:
            asset_hash: (str) asset to lookup, example would be 'c56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b'
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use

        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_ASSET_STATE, params=[asset_hash], id=id, endpoint=endpoint)

    def get_best_blockhash(self, id=None, endpoint=None):
        """
        Get the hash of the highest block
        Args:
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_BEST_BLOCK_HASH, id=id, endpoint=endpoint)

    def get_block(self, height_or_hash, id=None, endpoint=None):
        """
        Look up a block by the height or hash of the block.
        Args:
            height_or_hash: (int or str) either the height of the desired block or its hash in the form '1e67372c158a4cfbb17b9ad3aaae77001a4247a00318e354c62e53b56af4006f'
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use

        Returns:
            block: a json object or the ``neorpc.Core.Block.Block`` object
        """
        return self._call_endpoint(GET_BLOCK, params=[height_or_hash, 1], id=id, endpoint=endpoint)

    def get_block_hash(self, height, id=None, endpoint=None):
        """
        Get hash of a block by its height
        Args:
            height: (int) height of the block to lookup
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use

        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_BLOCK_HASH, params=[height], id=id, endpoint=endpoint)

    def get_block_sysfee(self, height, id=None, endpoint=None):
        """
        Get the system fee of a block by height.  This is used in calculating gas claims
        Args:
            height: (int) height of the block to lookup
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use

        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_BLOCK_SYS_FEE, params=[height], id=id, endpoint=endpoint)

    def get_connection_count(self, id=None, endpoint=None):
        """
        Gets the number of nodes connected to the endpoint
        Args:
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_CONNECTION_COUNT, params=[], id=id, endpoint=endpoint)

    def get_contract_state(self, contract_hash, id=None, endpoint=None):
        """
        Get a contract state object by its hash
        Args:
            contract_hash: (str) the hash of the contract to lookup, for example 'd7678dd97c000be3f33e9362e673101bac4ca654'
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_CONTRACT_STATE, params=[contract_hash], id=id, endpoint=endpoint)

    def get_raw_mempool(self, id=None, endpoint=None):
        """
        Returns the tx that are in the memorypool of the endpoint
        Args:
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_RAW_MEMPOOL, params=[], id=id, endpoint=endpoint)

    def get_transaction(self, tx_hash, id=None, endpoint=None):
        """
        Look up a transaction by hash.
        Args:
            tx_hash: (str) hash in the form '58c634f81fbd4ae2733d7e3930a9849021840fc19dc6af064d6f2812a333f91d'
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use

        Returns:
            json: the transaction as a json object
        """
        return self._call_endpoint(GET_RAW_TRANSACTION, params=[tx_hash, 1], id=id, endpoint=endpoint)

    def get_storage(self, contract_hash, storage_key, id=None, endpoint=None):
        """
        Returns a storage item of a specified contract
        Args:
            contract_hash: (str) hash of the contract to lookup, for example 'd7678dd97c000be3f33e9362e673101bac4ca654'
            storage_key: (str) storage key to lookup, for example 'totalSupply'
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            bytearray: bytearray value of the storage item
        """
        result = self._call_endpoint(GET_STORAGE, params=[contract_hash, binascii.hexlify(storage_key.encode('utf-8')).decode('utf-8')], id=id, endpoint=endpoint)
        try:

            return bytearray(binascii.unhexlify(result.encode('utf-8')))
        except Exception as e:
            print("could not decode result %s " % e)
        return None

    def get_tx_out(self, tx_hash, vout_id, id=None, endpoint=None):
        """
        Gets a transaction output by specified transaction hash and output index
        Args:
            tx_hash: (str) hash in the form '58c634f81fbd4ae2733d7e3930a9849021840fc19dc6af064d6f2812a333f91d'
            vout_id: (int) index of the transaction output in the transaction
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_TX_OUT, params=[tx_hash, vout_id], id=id, endpoint=endpoint)

    def invoke_contract(self, contract_hash, params, id=None, endpoint=None):
        """
        Invokes a contract
        Args:
            contract_hash: (str) hash of the contract, for example 'd7678dd97c000be3f33e9362e673101bac4ca654'
            params: (list) a list of json ContractParameters to pass along with the invocation, example [{'type':7,'value':'symbol'},{'type':16, 'value':[]}]
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(INVOKE, params=[contract_hash, params], id=id, endpoint=endpoint)

    def invoke_contract_fn(self, contract_hash, operation, params=None, id=None, endpoint=None):
        """
        Invokes a contract
        Args:
            contract_hash: (str) hash of the contract, for example 'd7678dd97c000be3f33e9362e673101bac4ca654'
            operation: (str) the operation to call on the contract
            params: (list) a list of json ContractParameters to pass along with the invocation, example [{'type':7,'value':'symbol'},{'type':16, 'value':[]}]
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(INVOKE_FUNCTION, params=[contract_hash, operation, params if params else []], id=id, endpoint=endpoint)

    def invoke_script(self, script, id=None, endpoint=None):
        """
        Invokes a script that has been assembled
        Args:
            script: (str) a hexlified string of a contract invocation script, example '00c10b746f74616c537570706c796754a64cac1b1073e662933ef3e30b007cd98d67d7'
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(INVOKE_SCRIPT, params=[script], id=id, endpoint=endpoint)

    def send_raw_tx(self, serialized_tx, id=None, endpoint=None):
        """
        Submits a serialized tx to the network
        Args:
            serialized_tx: (str) a hexlified string of a transaction
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use
        Returns:
            bool: whether the tx was accepted or not
        """
        return self._call_endpoint(SEND_TX, params=[serialized_tx], id=id, endpoint=endpoint)

    def validate_addr(self, address, id=None, endpoint=None):
        """
        returns whether or not addr string is valid

        Args:
            address: (str) address to lookup ( in format 'AXjaFSP23Jkbe6Pk9pPGT6NBDs1HVdqaXK')
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use

        Returns:
            json object of the result or the error encountered in the RPC call

        """
        return self._call_endpoint(VALIDATE_ADDR, params=[address], id=id, endpoint=endpoint)

    def get_peers(self, id=None, endpoint=None):
        """
        Get the current peers of a remote node
        Args:
            id: (int, optional) id to use for response tracking
            endpoint: (RPCEndpoint, optional) endpoint to specify to use

        Returns:
            json object of the result or the error encountered in the RPC call
        """
        return self._call_endpoint(GET_PEERS, id=id, endpoint=endpoint)

# Not all endpoints currently implement this method
#    def get_version(self, id=None, endpoint=None):
#        """
#        Get the current version of the endpoint
#        Args:
#            id: (int, optional) id to use for response tracking
#            endpoint: (RPCEndpoint, optional) endpoint to specify to use
#        Returns:
#            json object of the result or the error encountered in the RPC call
#        """
#        return self._call_endpoint(GET_VERSION, id=id, endpoint=endpoint)

    def __init__(self, config=None, setup=False):

        if config:
            self._settings = config

        self._build_addr()

        if setup:
            self.setup_endpoints()

    def setup_endpoints(self):
        self._build_addr()
        [endpoint.setup() for endpoint in self._addr_list]

    def _call_endpoint(self, method, params=None, id=None, endpoint=None):
        payload = self._build_payload(method, params, id)
        endpoint = self.default_endpoint if endpoint is None else endpoint
        try:
            response = requests.post(endpoint.addr, json=payload, timeout=TIMEOUT)
            response.raise_for_status()
            if response.status_code == 200:
                if 'result' in response.json():
                    return response.json()['result']
            return response.json()
        except Exception as e:
            print("Could not call method %s with endpoint: %s : %s " % (method, endpoint.addr, e))
        return None

    def _build_addr(self):
        self._addr_list = [RPCEndpoint(self, addr) for addr in self._settings.RPC_LIST]

    def _build_payload(self, method, params, id):

        id = self.id_counter if id is None else id
        self.id_counter += 1

        params = [] if params is None else params

        rpc_version = "2.0"

        return {'jsonrpc': rpc_version, 'method': method, 'params': params, 'id': id}


# methods that read data
GET_ACCOUNT_STATE = 'getaccountstate'
GET_ASSET_STATE = 'getassetstate'
GET_BEST_BLOCK_HASH = 'getbestblockhash'
GET_BLOCK = 'getblock'
GET_BLOCK_COUNT = 'getblockcount'
GET_BLOCK_HASH = 'getblockhash'
GET_BLOCK_SYS_FEE = 'getblocksysfee'
GET_CONNECTION_COUNT = 'getconnectioncount'
GET_CONTRACT_STATE = 'getcontractstate'
GET_RAW_MEMPOOL = 'getrawmempool'
GET_RAW_TRANSACTION = 'getrawtransaction'
GET_STORAGE = 'getstorage'
GET_TX_OUT = 'gettxout'
GET_PEERS = 'getpeers'
GET_VERSION = 'getversion'

# invocation related methods
INVOKE = 'invoke'
INVOKE_FUNCTION = 'invokefunction'
INVOKE_SCRIPT = 'invokescript'

# send
SEND_TX = 'sendrawtransaction'
SUBMIT_BLOCK = 'submitblock'

# validate
VALIDATE_ADDR = 'validateaddress'


TIMEOUT = 10


class RPCEndpoint():
    addr = None
    height = None
    client = None
    status = None
    elapsed = None

    def __init__(self, client, address):
        self.client = client
        self.addr = address

    def setup(self):

        response = requests.post(self.addr, json={'jsonrpc': '2.0', 'method': GET_BLOCK_COUNT, 'params': [], 'id': 1})
        self.update_endpoint_details(response)

        if response.status_code == 200:
            json = response.json()
            self.height = int(json['result'])

    def update_endpoint_details(self, response):

        self.status = response.status_code
        self.elapsed = response.elapsed.microseconds

    def _compare(self, other):
        if self.status != 200:
            return -1
        elif other.status != 200:
            return 1

        if self.height == other.height:
            if other.elapsed > 0 and self.elapsed > 0:
                if self.elapsed > other.elapsed:
                    return 1
                else:
                    return -1
        else:

            if self.height < other.height:
                return 1
            else:
                return -1

        return 0

    def __eq__(self, other):
        return self.addr == other.addr

    def __lt__(self, other):
        return self._compare(other) < 0

    def __gt__(self, other):
        return self._compare(other) > 0

    def __le__(self, other):
        return self._compare(other) <= 0

    def __ge__(self, other):
        return self._compare(other) >= 0

    def __str__(self):
        return "[%s] %s %s %s  " % (self.addr, self.status, self.height, self.elapsed)
